librad/src/net/quic.rs

```diff
@@ -15,32 +15,325 @@
+    pub async fn bind<'a>(
+        local_key: &device::Key,
+        listen_addr: SocketAddr,
+    ) -> Result<BoundEndpoint<'a>> {
+        let (endpoint, incoming) = make_endpoint(local_key, listen_addr).await?;
+        let endpoint = Endpoint {
+            peer_id: PeerId::from(local_key),
+            endpoint,
+        };
+        let incoming = incoming
+            .filter_map(|connecting| async move { connecting.await.ok().map(new_connection) })
+            .boxed();
+
+        Ok(BoundEndpoint { endpoint, incoming })
+    }
+
+    pub async fn connect<'a>(
+        &mut self,
+        peer: &PeerId,
+        addr: &SocketAddr,
+    ) -> Result<(Connection, BoxStream<'a, Result<Stream>>)> {
+        let conn = self
+            .endpoint
+            .connect(addr, peer.as_dns_name().as_ref().into())?
+            .await?;
+
+        Ok(new_connection(conn))
+    }
+}
+
+impl LocalInfo for Endpoint {
+    fn local_peer_id(&self) -> &PeerId {
+        &self.peer_id
+    }
+
+    fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.endpoint.local_addr()
+    }
+}
+
+pub type Incoming<'a> = BoxStream<'a, Result<Stream>>;
+
+pub struct BoundEndpoint<'a> {
+    pub endpoint: Endpoint,
+    pub incoming: BoxStream<'a, (Connection, Incoming<'a>)>,
+}
+
+impl<'a> LocalInfo for BoundEndpoint<'a> {
+    fn local_peer_id(&self) -> &PeerId {
+        self.endpoint.local_peer_id()
+    }
+
+    fn local_addr(&self) -> io::Result<SocketAddr> {
+        self.endpoint.local_addr()
+    }
+}
+
+fn new_connection<'a>(
+    NewConnection {
+        connection,
+        bi_streams,
+        ..
+    }: NewConnection,
+) -> (Connection, Incoming<'a>) {
+    let peer_id = tls::extract_peer_id(
+        connection
+            .authentication_data()
+            .peer_certificates
+            .expect("Certificates must be presented. qed")
+            .iter()
+            .next()
+            .expect("One certificate must have been presented. qed")
+            .as_ref(),
+    )
+    .expect("TLS layer ensures the cert contains a PeerId. qed");
+
+    let conn = Connection::new(&peer_id, connection);
+
+    (
+        conn.clone(),
+        Box::pin(
+            bi_streams
+                .map_ok(move |(send, recv)| Stream {
+                    conn: conn.clone(),
+                    send: SendStream {
+                        conn: conn.clone(),
+                        send,
+                    },
+                    recv: RecvStream {
+                        conn: conn.clone(),
+                        recv,
+                    },
+                })
+                .map_err(|e| e.into()),
+        ),
+    )
+}
+
+#[derive(Debug, Error)]
+pub enum ConnectionError {
+    #[error(transparent)]
+    Connection(#[from] quinn::ConnectionError),
+}
+
+#[derive(Clone)]
+pub struct Connection {
+    peer: PeerId,
+    conn: quinn::Connection,
+}
+
+impl Connection {
+    pub fn new(peer: &PeerId, conn: quinn::Connection) -> Self {
```

Comment:
`PeerId`s actually have to be moved all the time, not least because async code requires moves all the time -- so I think this convention is rather meaningless here, because we'd still clone inside the method body. 

I'm thinking that `PeerId` should just be `Copy` (it's 32 bytes), and can only speculate why the sodium wrapper for the pubkey isn't. The only worry is that we will have _a lot_ of those values, so perhaps some tweaks can be made to hold references in some places. In any case, material for another PR...

Reply:
Cool, was just curious :) Thought that might have been the case :ok_hand: